src/ray/core_worker/core_worker.cc

```diff
@@ -1803,6 +1804,15 @@ void CoreWorker::HandlePushTask(const rpc::PushTaskRequest &request,
   }
 
   task_queue_length_ += 1;
+
+  const TaskSpecification task_spec(request.task_spec());
+  {
+    absl::MutexLock lock(&direct_task_receiver_->mu_);
+    RAY_LOG(DEBUG) << "Received task " << task_spec.TaskId()
+                   << ". Adding it to the tasks_received_ queue!";
+    direct_task_receiver_->tasks_received_.emplace(task_spec.TaskId(), task_spec);
```

Comment:
> Could you also add a high-level description of the algorithm in the PR text? It would be nice to see a comparison of the flow diagram between the previous code and the new code, e.g., when are new workers requested?
> 
> I also had a question about when workers are returned if there are no tasks left to steal. Maybe this is already covered, but I didn't read the code in direct_task_transport.cc very carefully :) For example, what happens in the following case?
> 
> 1. Submit two tasks to worker A.
> 2. Request worker B since there is a task to steal.
> 3. Worker B is granted. Try to steal from worker A
> 4. Worker A finishes both tasks, responds 0 to the steal request.
> 
> When do we return worker B in this case?

Sure!

In the situation you mention here, worker B's `StealWork` callback, upon receiving a negative response from worker A (the victim), would call the `ReturnWorker` function to return worker B to the Raylet.